internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,155 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package speakerlist
+
+import (
+	"crypto/sha256"
+	golog "log"
+	"strconv"
+	"time"
+
+	"go.universe.tf/metallb/internal/k8s"
+
+	gokitlog "github.com/go-kit/kit/log"
+	"github.com/hashicorp/memberlist"
+)
+
+// SpeakerList represents a list of healthy speakers.
+type SpeakerList struct {
+	l         gokitlog.Logger
+	client    *k8s.Client
+	stopCh    chan struct{}
+	namespace string
+	labels    string
+
+	// The following fields are nil when memberlist is disabled.
+	mlEventCh chan memberlist.NodeEvent
+	mList     *memberlist.Memberlist
+}
+
+// New creates a new SpeakerList and returns a pointer to it.
+func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace, labels string, stopCh chan struct{}) (*SpeakerList, error) {
```

Comment:
I don't think I agree with this. At least with how we have things now.

The channel is created out of here and is passed to `internal/k8s` too. Just closing that channel signals all (speakerlist, intarnal/k8s, etc.) to close. Please note that the close method on a channel is sent to everyone receiving on that channel.

If the new is going to create a channel here instead, then we would need to keep track of several channels and close them all in the calling places. That seems quite impractical, IMHO, and not a minor change.

We can revisit this when doing a new minor release, IMHO (please create an issue if you think it is important to not forget about this. IMHO, if this is important it will be obvious when improving the code).

Marking as resolved for those reasons, feel free to re-open if I missed something :)

Reply:
All good, I guess I looked over the fact that we're using this channel with the fan-out pattern.